Add JSON_MODIFY()

anujsarda3 opened the pull request:
This function is used to update the value of the property in JSON string. Test cases are also added for the implementation. 

Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

### Issues Resolved

BABEL-937

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

Initial commit to add JSON_MODIFY feature and the test file BABEL-937

Task: BABEL-937
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Completed the functionality of  JSON_MODIFY and added test cases to BABEL-937.sql

Updated the test cases and made edits in the JSON_MODIFY function
implementation.

Task: BABEL-937
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

Sairakan commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,143 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+LANGUAGE plpgsql;
+
+--Function to set the append_modifier and create_if_missing flags
+CREATE TYPE falgs_set AS (create_if_missing BOOL, append_modifier BOOL);
```
typo here; should be `flags_set`, not `falgs_set`

Sairakan replied:
Also, the name of the datatype is not very descriptive - what if someone else wants to create a flag type? It should be more specific to its use, like `sys.json_modify_flags`. Also, make sure to declare this in the sys schema.

anujsarda3 replied:
Have made the changes and named the datatype sys.json_modify_flags

Sairakan commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,143 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
 AS 'babelfishpg_tsql', 'tsql_json_query' LANGUAGE C IMMUTABLE PARALLEL SAFE;
 
+-- Function to convert the path format into expected jsonb_set path format using Regular Expression
+CREATE OR REPLACE FUNCTION sys.regex_replace(json_path TEXT)
```
maybe a more descriptive name than just `regex_replace()`? Something that describes what the function does instead of how it works, so something like `convert_jsonpath()` or `convert_jsonpath_to_jsonb_path()`

anujsarda3 replied:
Have changed the function name to make it more specific sys.convert_jsonpath_to_jsonb_path

anujsarda3 committed:
Added the expected output file BABEL-937.out

TASK: BABEL-937
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

Sairakan commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,143 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+--Function to set the append_modifier and create_if_missing flags
+CREATE TYPE falgs_set AS (create_if_missing BOOL, append_modifier BOOL);
+
+CREATE OR REPLACE FUNCTION sys.set_flags(path_json TEXT)
```
Similar to above, change to `sys.set_json_modify_flags` to be more descriptive and specific

Sairakan commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,143 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+$BODY$
+DECLARE
+    json_path TEXT;
+    result falgs_set;
```
result is not a good variable name for this - makes it seem like we're saving the result in this variable.

anujsarda3 replied:
Have changed the variable name to mode_flags

anujsarda3 committed:
Code edits in JSON_MODIFY and added test cases

Made code edits as mentioned in the pull request. Added some test cases in the BABEL-937.sql file and expected output.

TASK: BABEL-937
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

Sairakan commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,143 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+            key_value = jsonb_path_query_first(json_string::JSONB,json_path::jsonpath); -- To get the value of the key
+            type_key_value = jsonb_typeof(key_value::JSONB);
+            IF type_key_value='array' THEN
+                len_array = jsonb_array_length(key_value::JSONB);
```
small nitpick: casting is expensive, so try to cast once and save the value in a variable for reuse instead of casting types multiple times

anujsarda3 replied:
Have made changes in the code to reduce the number of cast done.

Sairakan commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,143 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+                len_array = jsonb_array_length(key_value::JSONB);
+                new_jsonb_path = CONCAT(TRIM('}' FROM new_jsonb_path),',',CAST(len_array AS TEXT),'}'); -- To change the path format into the required jsonb_insert path format
+                IF new_value IS NULL THEN
+                    RETURN jsonb_insert(json_string::JSONB,new_jsonb_path::TEXT[],'null');
```
Can you combine all of these return calls into one return call at the bottom of the main IF block? And instead set some variable to pass into the return statement

Sairakan commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,143 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+    ELSE --When no append modifier is present
+        IF new_value IS NOT NULL THEN
+            IF key_exist OR mode_flags.create_if_missing='TRUE' THEN
+                RETURN jsonb_set_lax(json_string::JSONB,new_jsonb_path::TEXT[],to_jsonb(new_value),mode_flags.create_if_missing);
```
similar comment to above re: return statement

anujsarda3 committed:
Update JSON_MODIFY Code

Changed the code to reduce the number of cast.

TASK: BABEL-937
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Update in JSON_MODIFY code

Made changes in the return statements. Stored in a variable and returned.

TASK: BABEL-937
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,146 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+    append_modifier BOOL:='FALSE';
+BEGIN
+    
+    word_count = LENGTH(path_json) - LENGTH(REPLACE(path_json, ' ', '')) + 1; --To calculate the number of words in the path_json
```
Why not just count the number of spaces? Also, what happens with JSON that has multiple spaces between words? What about CR/LF?

nasbyj replied:
Also, since there's several code paths that need to split parts of this out, it might be better to just do `string_to_array()` or `regexp_split_to_array()` once and be done with it.

anujsarda3 replied:
Yes, you are right. It will be better to use regexp_split_to_array() instead of this. I will make the changes and implement it using regexp_split_to_array().

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,146 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+            create_if_missing:='FALSE';
+        END IF;
+    END IF;
+    SELECT create_if_missing,append_modifier INTO result_mode_flags.create_if_missing, result_mode_flags.append_modifier;
```
FWIW I think it's clearer to just use `:=` instead of `SELECT INTO`, especially since `SELECT INTO` can be confused with the engine syntax.

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,146 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+    type_key_value TEXT;
+    key_exist BOOL;
+    key_value JSONB;
+    json_expression JSONB;
```
to_jsonb() is pointless after doing `expression::jsonb`. Just do 

`json_expression jsonb := expression::jsonb;`

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,146 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+    word_count = LENGTH(path_json) - LENGTH(REPLACE(path_json, ' ', '')) + 1; --To calculate the number of words in the path_json
+    
+    -- This if else block is added to set the create_if_missing and append_modifier flags
+    IF word_count=1 THEN --The path_json has only 1 word
```
This comment is really just noise... it's clear from `IF word_count=1` that this is the one-word case.

anujsarda3 committed:
Update JSON_MODIFY code

Implemented the convert and set_flags in the json_modify function.

TASK: BABEL-937
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,119 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+    append_modifier BOOL:='FALSE';
+    key_exist BOOL;
+    key_value JSONB;
+    json_expression JSONB;
```
The `to_jsonb()` below is pointless after you've casted to jsonb. That whole line should just be removed by doing the assignment in the DECLARE section. While you're at it, you could also declare the variable as a constant:

```
json_expression CONSTANT jsonb := expression::jsonb;
```

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,119 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
 AS 'babelfishpg_tsql', 'tsql_json_query' LANGUAGE C IMMUTABLE PARALLEL SAFE;
 
+--JSON_MODIFY
+CREATE OR REPLACE FUNCTION sys.json_modify(in expression NVARCHAR,in path_json TEXT, in new_value TEXT)
```
This function needs a comment block explaining what the overall big picture here is. It's extremely difficult to follow any of this code without that.

anujsarda3 replied:
Will add more explanation about the implementation in my next commit.

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,119 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+    new_jsonb_path = CONCAT('{',json_path_convert,'}'); -- Final required format of path by jsonb_set
+    
+    
+    key_exist = jsonb_path_exists(json_expression,json_path::jsonpath); -- TO check if key exist in the given path
```
*Why* do we care if the key exists?

nasbyj replied:
Also, nitpick: should be `key_exists`.

anujsarda3 replied:
For append modifier:
If the key exists does not exists then we need to convert the new_value as array and then add it in the expression or else we do not need to convert the new_value to array_type and also we only check the value of the key only if the key_exists.

If append modifier is not present:
for strict mode the function works in different way if the key_exists or not. It returns an error if key exists and returns a new value in case key_exists.

Overall key_exists makes it easy to divide working of the function in different cases.

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,119 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+    
+    
+    key_exist = jsonb_path_exists(json_expression,json_path::jsonpath); -- TO check if key exist in the given path
+    word_count = LENGTH(path_json) - LENGTH(REPLACE(path_json, ' ', '')) + 1; --To calculate the number of words in the path_json
```
Given that several cases below care about the first 2 arguments, it might be better to just break the path apart using `string_to_array()` or `regexp_split_to_array()`. Also, there's a serious bug here. Consider this perfectly valid JSON:

```
{
    "key": "value",
    "other key":     NULL
)
```

anujsarda3 replied:
The path that I am splitting is the input json_path and not the input expression here. The syntax of the input json path is "`[append] [ lax | strict ] $.<json path>`".

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,119 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+    word_count = LENGTH(path_json) - LENGTH(REPLACE(path_json, ' ', '')) + 1; --To calculate the number of words in the path_json
+    
+
+    -- This if else block is added to set the create_if_missing and append_modifier flags
```
What the block is doing is already pretty obvious; what's absolutely not obvious is *why* it's doing it.

anujsarda3 replied:
will add comment to explain the use of setting up the flags in the next commit.

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,119 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+
+    -- This if else block is added to set the create_if_missing and append_modifier flags
+    IF word_count=1 THEN --The path_json has only 1 word
+        create_if_missing='TRUE';
```
Please don't wrap boolean values being assigned to a boolean in quotes. It's more to read and more work for the engine as well.

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,119 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+        append_modifier='FALSE';
+    ELSIF word_count=2 THEN --The path_json has 2 words
+        mode = SPLIT_PART(path_json COLLATE sql_latin1_general_cp1_cs_as, ' ',1);
+        IF mode='append' THEN
```
Do we *really* want this comparison to be case-sensitive?

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,119 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+            create_if_missing:='FALSE';
+        END IF;
+    ELSIF word_count=3 THEN --The path_json has 3 words
+        modifier = SPLIT_PART(path_json COLLATE sql_latin1_general_cp1_cs_as, ' ',1);
```
In the stanza above we call the 1st word "mode", but here we call it "modifier" even though it's checking for append, which is very confusing.

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,119 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+        END IF;
+        IF mode='strict' THEN
+            create_if_missing:='FALSE';
+        END IF;
```
What about the case of "strict append"? Is that definitely considered to be invalid? If so, maybe we should throw an error? (What does MSSQL do here?) Obviously if it is valid then we need to support it.

anujsarda3 replied:
So the syntax of the input path is defined as below:
[append] [ lax | strict ] $.<json path>

So even if the path is defined in the way "strict append", MSSQL throws an error 'JSON path is not properly formatted.' I have implemented this and can be seen in the next commit.

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,119 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+
+    --This if else block is to call the jsonb_set function based on the create_if_missing and append_modifier flags
+    
+    IF append_modifier THEN --When append modifier is present
```
Comment isn't useful; just remove it.

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,119 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+                result_json = jsonb_insert(json_expression,new_jsonb_path,to_jsonb(array_agg(new_value))); -- array_agg is used to convert the new_value text into array format as we append functionality is being used
+            END IF;
+        END IF;
+    ELSE --When no append modifier is present
```
*This* comment actually is useful since this ELSE is 26 lines away from it's IF.

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,119 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+                END IF;
+            END IF;
+        END IF;
+    END IF;
```
It'd be worth adding a comment to the effect of `-- IF append_modifier` here.

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,119 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+    IF append_modifier THEN --When append modifier is present
+        IF key_exist THEN
+            key_value = jsonb_path_query_first(json_expression,json_path::jsonpath); -- To get the value of the key
+            type_key_value = jsonb_typeof(key_value);
```
Nit: `key_value_type`, or maybe `key_value__type`. Or maybe `type_of_key_value`.

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,119 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+            type_key_value = jsonb_typeof(key_value);
+            IF type_key_value='array' THEN
+                len_array = jsonb_array_length(key_value);
+                new_jsonb_path = CONCAT(TRIM('}' FROM new_jsonb_path::TEXT),',',CAST(len_array AS TEXT),'}'); -- To change the path format into the required jsonb_insert path format
```
FWIW, `format()` might be clearer here than `concat()`. If you want to use concat I suggest splitting this call across several lines, ie:

```
new_jsonb_path = concat(
  '{'
  , trim(...)
  , '}'
)
```

nasbyj replied:
Also, the comment would be MUCH more useful if it gave an example of the two formats.

anujsarda3 replied:
have used format instead of concat

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,119 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+                len_array = jsonb_array_length(key_value);
+                new_jsonb_path = CONCAT(TRIM('}' FROM new_jsonb_path::TEXT),',',CAST(len_array AS TEXT),'}'); -- To change the path format into the required jsonb_insert path format
+                IF new_value IS NULL THEN
+                    result_json = jsonb_insert(json_expression,new_jsonb_path,'null');
```
Please add a comment that we can't simply do "to_jsonb(coalesce(new_value, 'null'))" here, because that won't result in a JSON NULL".

nasbyj commented on test/JDBC/expected/BABEL-937.out:
```diff
@@ -0,0 +1,210 @@
+-- Not Append
+--NOT NULL, LAX, KP  
+SELECT JSON_MODIFY('{"name":"John","skills":["C#","SQL"]}','$.name','Mike');
```
Please add some examples that use fully formatted JSON (such as what you'd get with `jsonb_pretty()`).

anujsarda3 replied:
Added the test cases.

nasbyj commented on test/JDBC/expected/BABEL-937.out:
```diff
@@ -0,0 +1,210 @@
+
+
+--NOT NULL, STRICT, KP
+SELECT JSON_MODIFY('{"name":"John","skills":["C#","SQL"]}','strict $.name','Mike');
```
Need to test upper/mixed case, as well as extra spaces, leading/trailing spaces etc. Even if those aren't supported we should know what error they're returning.

anujsarda3 replied:
Have tested the upper/mixed case and also leading/trailing spaces and have handled them in the implementation. Also added the test cases in the BABLE-937.sql file.

anujsarda3 committed:
Updated JSON_MODIFY code and BABEL-937.sql file

1. Updated the implementation of word count.
2. Took care of wrong json_path inputs.
2. Added comments to explain implementation details.
3. Minor code changes in variable declarations and change of functions.
4. Added test cases for case-sensitive and extra space handling.

TASK: BABEL-937
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Adding test files for BABEL-937

Added 3 files prepare,verify and cleanup according to the test format.

TASK: BABEL-937
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Updates in test files for BABEL-937

updates in verify, prepare and cleanup test file and added the .out files in the expected folder

TASK: BABEL-937
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Update in upgrade file for JSON_MODIFY

Added the function definition in upgrade file and also added the
function definition in the test output.

TASK: BABEL-937
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Redo the changes made in expected_dependency

Redo the changes made in expected_dependency.out file

TASK: BABEL-937
Signed-off-by: Anuj Sarda<sardanuj@amazon.com>

anujsarda3 committed:
Update in the schedule file for JSON_MODIFY

Removed BABEL-937 from schedule file for all versions

TASK: BABEL-937
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Change of type NVARCHAR in JSON_MODIFY

Changed the type NVARCHAR to sys.NVARCHAR.
Removed BABEL-937 from previous versions.

TASK: BABEL-937
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

2jungkook commented on test/JDBC/input/BABEL-937.sql:
```diff
@@ -0,0 +1,124 @@
+-- NOT APPEND
+--NOT NULL, LAX, KP  
+SELECT JSON_MODIFY('{"name":"John","skills":["C#","SQL"]}','$.name','Mike');
```
BABEL-937-vu tests are executed in both normal and upgrade tests. So let's move all test cases in this file into BABEL-937-vu. In the upgrade tests, please also use views instead of raw select statements. You can use a view to test multiple select statements, e.g.,
```
create view babel_937_vu_v2 as
select JSON_MODIFY(...),
       JSON_MODIFY(...),
       ...
```

2jungkook commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,146 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+3) To implement the main logic of the JSON_MODIFY function by dividing it into 8 different cases.
+
+*/
+CREATE OR REPLACE FUNCTION sys.json_modify(in expression sys.NVARCHAR,in path_json TEXT, in new_value TEXT)
```
Are input (nvarchar, text, text) and output (nvarchar) data types consistent with T-SQL's?

anujsarda3 replied:
Yes, they are consistent with T-SQL's datatypes.

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,146 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+    ELSIF word_count=2 THEN 
+        json_path = path_split_array[2];
+        comparison_string = path_split_array[1]; -- append or lax/strict mode
+        IF comparison_string = 'append' THEN
```
Are we intentionally making this a case-sensitive comparison?

anujsarda3 replied:
Yes, because it is a case-sensitive comparison. Otherwise it throws an error.

nasbyj replied:
It'd be nice to mention that in a comment then (and I hope there are test cases for this as well.

anujsarda3 replied:
Yes, there are test cases for case-sensitive comparison.

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,146 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+        ELSIF comparison_string = 'lax' THEN
+            create_if_missing:=TRUE;
+        ELSE
+            RAISE EXCEPTION 'JSON path is not properly formatted.';
```
Please review https://www.postgresql.org/docs/current/error-style-guide.html; in particular, error messages are not sentences, so the trailing `.` should go.

Need to use an explicit SQL state here. It would also be good to provide a DETAIL message showing what the problem is.

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,146 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+    END IF;
+
+    -- To convert input jsonpath to the required jsonb_path format
+    json_path_convert = regexp_replace(json_path, '\$\.|]|\$\[' , '' , 'ig'); -- To remove the "$." and "]" sign from the string 
```
nit: s/remove the/remove/

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,146 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+
+    -- To convert input jsonpath to the required jsonb_path format
+    json_path_convert = regexp_replace(json_path, '\$\.|]|\$\[' , '' , 'ig'); -- To remove the "$." and "]" sign from the string 
+    json_path_convert = regexp_replace(json_path_convert, '\.|\[' , ',' , 'ig'); -- To replace the "." and "[" with the "," to change into required format
```
nit: s/replace the/replace/, s/with the/with/

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,146 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+                END IF;
+            ELSE
+                IF NOT create_if_missing THEN
+                    RAISE EXCEPTION 'Array cannot be found in the specified JSON path.';
```
ERRMSG shouldn't be a sentence. Need SQLSTATE. Would be good to provide the user with a DETAIL message as well (which should be a sentence), as well as a HINT about how they could specify to create the key if it doesn't exist.

anujsarda3 replied:
Are you suggesting to display the SQL state error code instead of the default ‘33557097’?

nasbyj commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2540,6 +2540,146 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+BEGIN
+    path_split_array = regexp_split_to_array(TRIM(path_json) COLLATE "C",'\s+');
+    word_count = array_length(path_split_array,1);
+    /* 
```
nit: It'd be better to follow community block-comment formatting standard:

```
/*
 * This if else block is added to set the create_if_missing and append_modifier flags.
 * These flags will be used to know the mode and if the optional modifier append is present in the input path_json.
 * It is necessary as postgres functions do not directly take append and lax/strict mode in the jsonb_path.  
 */

anujsarda3 committed:
Updated JSON_MODIFY implementation and test cases

- Provided SQLSTATE to the error messages
- Created view to run multiple select statements
- Removed the file BABEL-937.sql and made changes in BABEL-937-vu files
- Followed the community block-comment formatting standard

TASK: BABEL-937
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Deleted BABEL-937 test file

- Deleted BABEL-937.sql and BABEL-937.out files

TASK: BABEL-937
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Updated expected_create.out file

Removed json_value line from the file expected_create.out

TASK: BABEL-937
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Updated expected_dependency.out file

Removed json_value line from the expected_dependancy.out

TASK: BABEL-937
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

2jungkook commented on test/JDBC/expected/BABEL-937-vu-prepare.out:
```diff
@@ -0,0 +1,107 @@
+SELECT JSON_MODIFY('{"name":"John","skills":["C#","SQL"]}','  aPpend    Strict    $.skills   ',NULL);
+go
+
+create view babel_937_case_sensitive_3 as
```
Why this is related to case sensitive?

anujsarda3 replied:
Resolved in latest commit

2jungkook commented on test/JDBC/expected/BABEL-937-vu-prepare.out:
```diff
@@ -0,0 +1,107 @@
+create view babel_937_not_append_kp as
```
From reviewer's perspective, it is not easy to know what `not_append_kp` means and what's the purpose of each test cases. Not only for this specific case but also for other test cases, please add some comments explaining which functionality is tested.

2jungkook commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2515,6 +2515,154 @@ CREATE OR REPLACE FUNCTION sys.json_query(json_string text, path text default '$
+            append_modifier = TRUE;
+        ELSIF comparison_string = 'strict' THEN
+            create_if_missing = FALSE;
+        ELSIF comparison_string = 'lax' THEN
```
Don't we need a test case for `lax`?

anujsarda3 replied:
As Lax mode is the default mode(when no mode is provided), so tests for lax mode are already covered in the test cases.

2jungkook replied:
Even if the behavior of lax and no mode is the same, there should be at least two test cases where lax is explicitly provided when word count is 2 or 3.

anujsarda3 replied:
Resolved in latest commit.

2jungkook commented on test/JDBC/expected/BABEL-937-vu-verify.out:
```diff
@@ -0,0 +1,136 @@
+go
+~~START~~
+nvarchar#!#nvarchar
+{"name": "Mike", "skills": ["C#", "SQL"], "surname": ["Smith"]}#!#{"k": [null], "name": "Mike", "skills": ["C#", "SQL"], "surname": "Smith"}
```
`surname` is added at the end, but `k` is added at the beginning. Is this expected?

anujsarda3 replied:
Yes, This is an expected and known behavior when using postgres functions.

2jungkook replied:
In that case, please add some comments in the test file that the result can be different between Babelfish and T-SQL.

anujsarda3 replied:
Resolved in latest commit.

anujsarda3 committed:
Updated Test cases files for JSON_MODIFY

- Added comments for different test cases
- changed the name of created view

TASK: BABEL-937
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Updated sys-all_sql_modules-vu-verify.out file

- Change the expected output file of sys-all_sql_modules-vu-verify

TASK: BABEL-937
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Updated the sys-all_sql_modules-vu-verify.out file

- Due to the fix from the commit f0f4d49d0c12f06da46c8c1c78d7d16bfcbf8b12,
  had to make changes in the file sys-all_sql_modules-vu-verify.out of
  expected folder.

TASK: BABEL-937
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Updayed the test cases for JSON_MODIFY

- Added explicit lax mode for some test cases
- Provided proper comments for some of the test cases

TASK: BABEL-937
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Merge branch 'BABEL_2_X_DEV' into jira-babel-937

-- Fix for the tests where 2 objects sharing the same object id.

anujsarda3 committed:
Merge branch 'BABEL_2_X_DEV' into jira-babel-937

nasbyj reviewed (dismissed):
Approved, though see my one comment about explicitly stating that we want to be case sensitive.

anujsarda3 committed:
Update in sys_functions.sql file

- Added comment for case-sensitive comparisons

TASK: BABEL-937
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Update in sys_functions.sql

- Made changes in the added comment for case-sensitive comparisons

TASK: BABEL-937
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Merge branch 'BABEL_2_X_DEV' into jira-babel-937

anujsarda3 committed:
Updated the schedule file

- updated the upgrade/latest/schedule file to resolve the conflicts

TASK: BABEL-937
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>
